samples/diagnostics/facts-and-assertions/Facts and Assertions.ipynb

```diff
@@ -0,0 +1,910 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Assertions: Conditions on machine state"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "By comparison, we sometimes want to check a condition on the state of our machine or of some qubits within a device. Unlike facts, conditions of this form can't be represented as functions, since a condition on a `Qubit` value may be true or false depending on the side effects accumulated in a program so far — not just depending on the `Qubit` itself.\n",
+    "\n",
+    "Thus, conditions on states need to represented as *operations* rather than as *functions*. Let's take a look at how we can use operations to validate that freshly allocated qubits start off in the $\\left|0\\right\\rangle$ state:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "application/json": "[\"CheckFreshQubitIsInZero\"]",
+      "text/html": [
+       "<ul><li>CheckFreshQubitIsInZero</li></ul>"
+      ],
+      "text/plain": [
+       "CheckFreshQubitIsInZero"
+      ]
+     },
+     "execution_count": 12,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "operation CheckFreshQubitIsInZero() : Unit {\n",
+    "    using (q = Qubit()) {\n",
+    "        AssertMeasurement([PauliZ], [q], Zero, \"qubit was not in |0⟩\");\n",
+    "    }\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "application/json": "{\"@type\":\"tuple\"}",
+      "text/plain": [
+       "()"
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "%simulate CheckFreshQubitIsInZero"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "When run on a simulator that supports it, [`AssertMeasurement`](https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.diagnostics.assertmeasurement) checks that **if** you were to measure `[q]` in the `[PauliZ]` basis, then it you would get the result `Zero` with **certianty**. Since the only state which always returns a measurement result of `Zero` when measured in the `PauliZ` basis is the $\\left|0\\right\\rangle$ state, passing the assertion tells us that our qubit is definitely in the $\\left|0\\right\\rangle$ state immediately afterwards.\n",
+    "\n",
+    "> ⚠ **NOTE**: This assertion critically depends on being able to violate the no-cloning theorem, since it checks a hypothetical measurement without actually performing that measurement.\n",
+    "> That said, since the assertion has no effect on the state of the target machine, it can be safely removed when run on simulators which don't support it, or on actual hardware that is bound by the no-cloning theorem.\n",
```

Comment:
I think *removed* implies that the user has to remove the line from the code when targeting hardware. But it should just be ignored by target machines.

Reply:
Good catch, yeah, let me fix that.